README.md
<table float="center">
  <tr>
    <td align="center" width=33%>
      <h1>v^SPEED Vario mini</h1>
      Open Source Paragliding Variometer.</br>
      Email <a href="mailto:vspeedvario@gmail.com">vSpeedVario@gmail.com</a> to purchase.</br>
      <h3>Usage and Instructions:</h3>
      <a href="#start-here">Start Here</a> </br>
      <a href="#how-to-change-custom-settings">Changing Settings</a></br>
      <a href="#bluetooth-operation">Bluetooth Operation</a> </br>
      <a href="#firmware-updates">Firmware Updates</a> </br>
    </td>
  </tr>
</table>
<table float="center">
  <tr>
    <td width=33% vAlign="top" align="center">
      <h2>v^SPEED VARIO mini</h2>
      <h3>$99</h3>
      <img src="https://github.com/glydrfreak/vSpeed-Vario/blob/master/Images/vspeedMini2.jpg" width=250> </br>
      <img src="https://github.com/glydrfreak/vSpeed-Vario/blob/master/Images/labels.jpg" width=350> </br>
      <h3>Email <a href="mailto:vspeedvario@gmail.com">vSpeedVario@gmail.com</a> to purchase</h3>
    </td>
  </tr>
</table>
<table float="center">
  <tr>
    <td width=66%> 
      <ul>
        <h3>SPECS:</h3>
        <li>Ultra-high precision barometric pressure sensor detects altitude changes as small as 10cm</li>
        <li>Measures altitude 84 times per second</li>
          (72 while transmitting bluetooth data)
        <li>Lowest priced Bluetooth Variometer on the market</li>
        <li>iPhone users: Fully compatible with <a href="#flyskyhy">FlySkyHy</a> app (recommended)</li>
        <li>Android users: <a href="#vspeed-vario">v^SPEED</a> app (experimental)</li>
          (Recommended use is with FlySkyHy for iPhone)
        <li>Fully compatiple with Arduino IDE to easily change custom settings</li>
          (Micro USB cable included)
        <li>Mounts on risers with velcro</li>
          (100lb test 0.55mm Dyneema Leash, and velcro wrap for riser included)
        <li>Tiny Size: 68x25x10mm (about the size of a man's thumb)</li>
        <li>3D printed case (any color available upon request)</li>
        <li>Power switch, and volume buttons</li>
        <li>150mAh lipo battery with microUSB charging port</li>
        <li>10hr battery life while streaming bluetooth data</li>
        <li>1.5hr charging time</li>
        <li>Sale price: $99 each</li>
      </ul>
    </td>
  </tr>
</table>


### ABOUT
This is an open source project, with all public files available at <a href="https://github.com/glydrfreak/vSpeed-Vario">the vSpeed-Vario repository</a>. The v^SPEED mini is based on PROTOTYPE6, which is found within that same repository. This variometer was designed to be compatible with the Arduino IDE for advanced usage, which allows users to modify the existing source code to their pleasure. Here, in the repository you are currently viewing, you will find the most up-to-date source code for use in the Arduino IDE. The hardware is based on another open source product: <a href="https://learn.adafruit.com/adafruit-feather-m0-bluefruit-le/downloads">Adafruit's Bluefruit Feather M0</a>. The v^SPEED mini is basically just a Bluefruit Feather M0 development board, but without the header pins broken out, and with the added variometer circuitry.


### INSTRUCTIONS
<a href="#start-here">Start Here</a> </br>
<a href="#how-to-change-custom-settings">Changing Settings</a></br>
<a href="#bluetooth-operation">Bluetooth Operation</a> </br>
<a href="#firmware-updates">Firmware Updates</a> </br>


### START HERE
#### Congrats on receiving your new v^SPEED Vario mini!
- Arduino IDE is an integrated development environment, which compiles and executes written code. All the prewritten code used by the vario is included in the repository on this website. If you desire to change any settings on the vario, you must obtain a copy of this code and edit it through the Arduino IDE and re-upload it to the vario. 
- Install Arduino IDE from <a href="https://www.arduino.cc/en/Main/Software">Arduino's website</a>
- Click <a href="https://github.com/glydrfreak/vSpeedVario-mini/archive/master.zip" raw=true>HERE</a> to download and extract the code files in this repository
- Open the `vSpeed_mini.ino` file, which should open with the Arduino IDE
- Tools ->  Board ->Boards Manager -> Search for: <i>Arduino SAMD Boards</i> -> Install
- File -> Preferences -> Next to "Additional Boards Manager URLs", type: `https://adafruit.com/package_adafruit_index.json`
  - (If a URL is already there, separate it with a comma and type the new URL next to it)
- Tools ->  Board ->Boards Manager -> Search for: <i>Adafruit SAMD Boards</i> -> Install
- Tools -> Board -> Select: Adafruit Feather M0
- Connect your v^SPEED mini to your computer
  - Some USB cables are "charge only" cables, which won't work for uploading code, so use the USB cable provided in the kit
  - The vario must be switched on
- Tools -> Port -> Select any port for now
- Sketch -> Upload
- Anytime you power on your vario, or upload a new sketch, you will hear two sequences of beeps:
  - The first beep(s) indicates the bluetooth mode.
    - No beep: Bluetooth disabled
    - One beep: vSpeed Android app mode
    - Two beeps: FlySkyHy iPhone app mode
    - Three beeps: XC Track Android app mode (not working yet)
  - The second series of beeps indicates the battery level.
    - Ascending: Battery is at least 66%
    - Steady: Battery is at least 33%
    - Descending: Battery is below 33%
- <b>You're ready to fly!</b>


### HOW TO CHANGE CUSTOM SETTINGS
- Open the `vSpeed_mini.ino` file, which should open with the Arduino IDE
- Open the `DEFAULT_SETTINGS.h` tab
- Change any of the following values according to your preference:
  - `ENABLE_CLIMB_BEEP`= <b>1</b>; [1]YES, [0]NO
  - `ENABLE_SINK_BEEP`= <b>0</b>; [1]YES, [0]NO
  - `ENABLE_SINK_ALARM`= <b>1</b>; [1]YES, [0]NO
  - `START_UP_VOLUME`= <b>10</b>; (0 TO 31)
  - `BEEP_TYPE`= <b>1</b>; Default [1]BASED_ON_VELOCITY, [2]BUFFERED_INCREMENTS
  - `CLIMB_BEEP_TRIGGER`= <b>1.0</b>; (Default: One beep for every 1.0 foot of altitude increase)
  - `SINK_BEEP_TRIGGER`= <b>-1.0</b>; (Default: Constant beep if sinking more than -1.0 ft/s)
  - `SINK_ALARM_TRIGGER`= <b>-20.0</b>; (Default: -20.0 ft/s)
  - `CLIMB_PITCH_MAX`= <b>600.0</b>; (Default: 600.0 Hz when climbing at ten ft/s)
  - `CLIMB_PITCH_MIN`= <b>400.0</b>; (Default: 400.0 Hz when climbing at one ft/s)
  - `SINK_PITCH_MAX`= <b>250.0</b>; (Default: 250.0 Hz when sinking at `SINK_BEEP_TRIGGER` ft/s)
  - `SINK_PITCH_MIN`= <b>150.0</b>; (Default: 150.0 Hz when sinking at ten ft/s less than `SINK_BEEP_TRIGGER` ft/s)
  - `SINK_ALARM_PITCH`= <b>1000.0</b>; (Default 1000.0 Hz)
  - `BLUETOOTH_MODE`= <b>2</b>; Default [0]DISABLE, [1]VSPEED_ANDROID, [2]FLYSKYHY_IOS (default), [3]XCTRACK_ANDROID (not working yet)
  - `ALTITUDE_UNITS`= <b>1</b>; Default [1]FEET, [2]METERS
  - `VELOCITY_UNITS`= <b>1</b>; Default [1]FEET_PER_SECOND, [2]METERS_PER_SECOND, [3] FEET_PER_MINUTE
- Tools -> Board -> Select: Adafruit Feather M0
- Connect your v^SPEED mini to your computer
  - Some USB cables are "charge only" cables, which won't work for uploading code, so use the USB cable provided in the kit
  - The vario must be switched on
- Tools -> Port -> Select the appropriate port 
  - The port should be automatically renamed to "COM# (Adafruit Feather M0)"
- Sketch -> Upload


### BLUETOOTH OPERATION
- v^SPEED mini is compatible with both Android and iPhone.
- You may change the bluetooth mode at startup:
  - Hold the volume down button while powering on the vario until it beeps to boot up into Android mode.
  - Hold the volume up button while powering on the vario until it beeps to boot up into iPhone mode.
  - Hold both the volume buttons while powering on the vario until it beeps to boot up into XC Track mode. (XC Track transmission is not working yet).
  - If no volume buttons are held down at startup, the default mode (as shipped) will be to boot up into FlySkyHy transmission mode.
    - You may change the default startup bluetooth mode by changing the `BLUETOOTH_MODE` in `DEFAULT_SETTINGS.h` so you don't need to hold down a volume button each time you switch on the vario.

#### Android Users--
- Download the <a href="https://play.google.com/store/apps/details?id=com.braedin.butler.vspeed.vario&hl=en_US">v^SPEED Vario app</a> from Google Play Store.
<table>
  <tr>
    <td width=33% vAlign="top" align="center">
      <h2>v^SPEED Vario</h2>
      <img src="https://github.com/glydrfreak/vSpeed-Vario/blob/master/Images/ANDROID.png" width=250>
    </td>
    <td width="66%"> 
      <ul>
        <h3>Android App--</h3>
        This is a modified version of Adafruit's open source "Bluefruit LE Connect" for Android. The source code can be found in the Android folder of <a href="https://github.com/glydrfreak/vSpeed-Vario">the vSpeed-Vario repository</a>. Now published to Google Play! </br>
  </br>
        This app will give you a better image of what your v^SPEED Vario is doing as it transmits live bluetooth data of barometric altitude, vertical speed, and battery level. Other features include implemented GPS for GPS altitude, ground speed, and heading. The implemented Audio beeps are helpful for listening to each foot of climb as well as a variable sink tone. An implemented accelerometer is for keeping track of max G-force during a hard turn or spiral. The visual display includes a line chart of vertical speed, as well as scrolling altitude numbers so you can see your precise vertical velocity. The climb and sink thresholds are adjustable within the app, which also change the thresholds on the external vario automatically.
      </ul>
    </td>
  </tr>
</table>

#### iPhone Users--
- Download <a href="https://itunes.apple.com/us/app/flyskyhy/id516879039?mt=8">Flyskyhy</a> from the App Store.
- While selecting the external vario model, choose `Skydrop`.
<table>
  <tr>
    <td width=33% vAlign="top" align="center">
      <h2>Flyskyhy</h2>
      <img src="https://github.com/glydrfreak/vSpeed-Vario/blob/master/Images/fsh2.jpg" width=250>
    </td>
    <td width=66%> 
      <ul>
        <h3>iOS App by Renevision--</h3>
        v^SPEED Vario is now compatible with the World's most popular Paragliding app! Flyskyhy is an unbeatable iPhone application when it comes to flying. v^Speed Vario is currently using the same protocol as SkyDrop variometers, so in the app when selecting your device, you must choose SkyDrop.  
      </ul>
    </td>
  </tr>
</table>


### FIRMWARE UPDATES
#### Source code last updated: `9/12/2020`
- For the most up-to-date source code for your v^SPEED mini, download and extract the <a href="https://github.com/glydrfreak/vSpeedVario-mini/archive/master.zip" raw=true>files</a> in this repository and replace the existing files in your `vSpeed_mini` folder. 
- You may want to keep your existing `DEFAULT_SETTINGS.h` file if you have made changes to it.
- To see the most recent date that your existing code was updated, see the first line of code in `vSpeed_mini.ino`.
- a `changeLog.txt` is provided in the `vSpeed_mini` folder.


### QUESTIONS / CONCERNS / ASSISTANCE / PURCHASE
Email vSpeedVario@gmail.com </br>


vSpeed_mini/changeLog.txt
9/12/2020 - Something broke in the original MS5611 library. The issue could not be identified, so the library was replaced with an alternative after being optimized for speed.

9/20/2018 - (Revert FILTER.h line 83 and vSpeed_mini.ino line 32.) The Temperature was not the problem. It was light hitting the sensor at just the right angle. Reverting changes to the software. Adding a fix for the hardware.

9/19/2018 - (FILTER.h line 83 and vSpeed_mini.ino line 32.) The temperature of the sun was having a very negative effect on the vario as the device passed back and forth from sun to shadow. It would cause the vertical speed calculations to go crazy, and sometimes even crash the device. At the cost of sample rate from 77 to 68, I applied a filter to the temperature data as a 10 second running average. Hopefully this provides the fix.

8/16/2018 - Got rid of the option to not measure battery level. It's just not nessesary to worry about with this prototype.

8/8/2018 - Limitted the transmit rate to Flyskyhy to 30 times per second instead of full speed at 69 times per second. The app seems like it might have had a hard time keeping up and would sometimes glitch with incorrect altitude and/or lift readings at least once every flight. Hopefully this fixes it.

8/2/2018 - Little bit of code cleanup by removing a redundant setting "ENABLE_BLUETOOTH" in "DEFAULT_SETTINGS.h", also removed the feature that disables BLE after two minutes of no connection. (The device will sometimes randomly disconnect from the v^SPEED Android app. Working on a fix.)

5/29/2018 - (vSpeed_mini.ino lines 147-175) fixed each case to be at each third of a battery life.

5/11/2018 - Initial release.

vSpeed_mini/BEEP.h
/*
	BEEP.h
	Variometer Beep Library written by Braedin and Paul Butler
	
*/

#ifndef BEEP_h
#define BEEP_h

#include <Arduino.h>


class BEEP{


private:	
  const int DURATION = 0;                    //NEW
  const int PITCH = 1;                       //NEW
  static const int beepBuffSize = 5;                //NEW
  float percentageOfCycleOn = 0.50;     //NEW
  bool allowInterrupt = true;           //NEW
  bool incrementDuration = true;        //NEW
  float beeps[2][beepBuffSize] = {};    //NEW
  int beepsWaitingToBeep = 0;           //NEW
  unsigned long beepMillis = 0;         //NEW
  bool needBeeps = false;               //NEW

	float altitudeTriggerMemory;
	unsigned long timeTriggerMemory;
	float beepDuration;
	int beepPitch;
  int sinkPitch;
  float leastClimb = 1.0;
  float mostClimb = 10.0;
  float leastSink = -1.0;
  float mostSink = -10.0;
  //#define dbg  false                // set true when debugging is needed
	int verticalTrigger = 1.0;		      // default feet
	int sinkAlarm = -1.0;		            // default feet per second
  int majorSinkAlarm = -20;           // default feet per second
	int sinkAlarmDuration = 500;	      // default milliseconds
	int sinkAlarmPitch = 250;	          // default Hz
	int sap = sinkAlarmPitch;	
  float climbDurationShort = 50.0;	  // default milliseconds
	float climbDurationLong = 300.0;	  // default milliseconds 
  float prevDur = climbDurationLong;
  bool enableSinkAlarm = true;
  bool enableSinkBeep = true;
  bool enableClimbBeep = true;
  
  

public:
  
  float pitchMax = 500.0;           // default Hz
  float pitchMin = 300.0;           // default Hz
  float sinkPitchMax = 250;//250
  float sinkPitchMin = 150;//150
  int beepWait = 0;
  int buzzerPin = A5; // default pin connected to the buzzer
  
  void basedOnVelocity(float currentAltitude, float velo, unsigned long currentTime);
  void basedOnAltitude(float currentAltitude, float velo, unsigned long currentTime);
  void durationIncrements(float currentAltitude, float velo, unsigned long currentTime);
  void bufferedDurationIncrements(float currentAltitude, float velo, unsigned long currentTime);
  
  
  // set buzzerPin
  void begin(int BP); 
  
  void setClimbThreshold(float VT){verticalTrigger = VT;}
  void setSinkAlarmThreshold(float SA){sinkAlarm = SA;}
  void setMajorSinkAlarm(float MSA){majorSinkAlarm = MSA;}
  void setClimbPitchMax(float HUBP){pitchMax = HUBP;}
  void setClimbPitchMin(float LUBP){pitchMin = LUBP;}
  void setSinkPitchMax(float SPX){sinkPitchMax = SPX;}
  void setSinkPitchMin(float SPN){sinkPitchMin = SPN;}
  void setSinkAlarmPitch(float SAP){sinkAlarmPitch = SAP; sap = SAP;}
  void setBeeps(bool ECB, bool ESB, bool ESA){
    enableClimbBeep = ECB;
    enableSinkBeep = ESB;
    enableSinkAlarm = ESA;
  }
  
};

#endif







vSpeed_mini/FILTER.h
/**************************************************************************/
/*
    @file     RUNNING_AVERAGE.h
    @author   Braedin Butler

    v^SPEED Vario for Paragliding:
    
    MS5611-01BA03_datasheet:
      http://www.te.com/commerce/DocumentDelivery/DDEController?Action=srchrtrv&DocNm=MS5611-01BA03&DocType=Data+Sheet&DocLang=English

    
    @section LICENSE
*/
/**************************************************************************/
#ifndef FILTER_H
#define FILTER_H

#include <Arduino.h>




class FILTER1{//PRESSURE:
 private:
  int samplesToAverage;
  static const int maxDataMemory = 200;   // make this number as small as possible without clipping the quantity of samples to be averaged
  bool f1 = true; 
  float previousData = 0;
  bool rejectBadData = false;     // If any new data is greater than rejectionThreshold away from the previous data, reject if true;
  int rejectionThreshold = 100; 
  
 public:
  //int AVERAGING_DURATION = 1000;     // Greater duration (max 1000 ms) means more samples averaged (see maxDataMemory set in FILTER.h)
  float DATA[maxDataMemory] = {};
  bool dbg = false;
	float RUNNING_AVERAGE(float newData, int samplesPerSec, float averagingDuration);
};




class FILTER2{//ALTITUDE:
 private:
  int samplesToAverage;
  static const int maxDataMemory = 200;   // make this number as small as possible without clipping the quantity of samples to be averaged
  bool f1 = true; 
  float previousData = 0;
  bool rejectBadData = false;     // If any new data is greater than rejectionThreshold away from the previous data, reject if true;
  int rejectionThreshold = 100; 
  
 public:
  //int AVERAGING_DURATION = 1000;     // Greater duration (max 1000 ms) means more samples averaged (see maxDataMemory set in FILTER.h)
  float DATA[maxDataMemory] = {};
  bool dbg = false;
  float RUNNING_AVERAGE(float newData, int samplesPerSec, float averagingDuration);
};




class FILTER3 {//VELOCITY:
private:
	int samplesToAverage;
	static const int maxDataMemory = 200;   // make this number as small as possible without clipping the quantity of samples to be averaged
	bool f1 = true;
	float previousData = 0;
  bool rejectBadData = false;     // If any new data is greater than rejectionThreshold away from the previous data, reject if true;
  int rejectionThreshold = 100; 

public:
	//int AVERAGING_DURATION = 1000;     // Greater duration (max 1000 ms) means more samples averaged (see maxDataMemory set in FILTER.h)
	float DATA[maxDataMemory] = {};
	bool dbg = false;
	float RUNNING_AVERAGE(float newData, int samplesPerSec, float averagingDuration);
};




class FILTER4 {//TEMPERATURE:
private:
  int samplesToAverage;
  static const int maxDataMemory = 1000;   // make this number as small as possible without clipping the quantity of samples to be averaged
  bool f1 = true;
  float previousData = 0;
  bool rejectBadData = false;     // If any new data is greater than rejectionThreshold away from the previous data, reject if true;
  int rejectionThreshold = 100; 

public:
  //int AVERAGING_DURATION = 1000;     // Greater duration (max 1000 ms) means more samples averaged (see maxDataMemory set in FILTER.h)
  float DATA[maxDataMemory] = {};
  bool dbg = false;
  float RUNNING_AVERAGE(float newData, int samplesPerSec, float averagingDuration);
};

#endif






vSpeed_mini/MS5611.h
//
//    FILE: MS5611.h
//  AUTHOR: Rob Tillaart
//          Erni - testing/fixes
// VERSION: 0.1.10
// PURPOSE: MS5611 Temperature & Atmospheric Pressure library for Arduino
//     URL:
//
// HISTORY:
// see MS5611.cpp file
//

#ifndef MS5611_h
#define MS5611_h

#if ARDUINO < 100
#error "VERSION NOT SUPPPORTED"
#else
#include <Arduino.h>
#endif

#define MS5611_LIB_VERSION (F("0.1.9"))

#define MS5611_READ_OK  0

class MS5611
{
public:
  explicit        MS5611(uint8_t CSpin);

  void            init();
  int             read(/*uint8_t bits = 8*/);
  inline int32_t  getTemperature()        const { return _temperature; };
  inline uint32_t getPressure()           const { return _pressure; };
  inline int      getLastResult()         const { return _result; };
  inline uint16_t getPromValue(uint8_t p) const { return C[p]; };

private:
  void     reset();
  void     convert(const uint8_t addr, uint8_t bits);
  void     wait(uint8_t OSR_mode);
  uint32_t readADC();
  uint16_t readProm(uint8_t reg);
  void     command(const uint8_t command);

  uint8_t  _cspin;
  int32_t  _temperature;
  uint32_t _pressure;
  int      _result;
  uint16_t C[8];
};
#endif

// END OF FILE


vSpeed_mini/MS5611.cpp
//
//    FILE: MS5611.cpp
//  AUTHOR: Rob Tillaart
//          Erni - testing/fixes
//          Roy van der Kraan - partial rewrite (mainly for SPI use)
//          Braedin Butler - tuned for vSpeed Vario
// VERSION: 0.1.10
// PURPOSE: MS5611 Temperature & Atmospheric Pressure library for Arduino
//     URL:
//
// POST FORK HISTOTY
// 0.1.10 2020-09-12 Tuned for speed and tuned OSRs to vSpeed Vario preferences.
// 0.1.9  2019-01-17 modified for SPI
//                   replaced all floating point with integer arithmatic
//
// PRE-FORK HISTORY:
// 0.1.8  fix #109 incorrect constants (thanks to flauth)
// 0.1.7  revert double to float (issue 33)
// 0.1.6  2015-07-12 refactor
// 0.1.05 moved 6 float multiplies to init  [adds ~70 bytes !!!]
//        moved the MS5611_LIB_VERSION to PROGMEM
// 0.1.04 changed float to double (for platforms which support it)
//        changed divisions in multiplications
//        fixed uint32_t readADC()
//        reduced size of C array by 1 float
//        added second order temperature compensation
// 0.1.03 changed math to float [test version]
// 0.1.02 fixed bug return value read()
//        fixed bug #bits D2
//        added MS5611_READ_OK
//        added inline getters for temp & pres & lastresult.
//        adjusted delay's based on datasheet
//        merged convert functions
//        fixed offset in readProm()
// 0.1.01 small refactoring
// 0.1.00 added temperature and Pressure code
// 0.0.00 initial version by Rob Tillaart (15-okt-2014)
//
// Released to the public domain
//

#include "MS5611.h"
#include <SPI.h>
SPISettings settingsA(1000000, MSBFIRST, SPI_MODE0);  // define SPI settings; limit SPI communication to 1MHz


/////////////////////////////////////////////////////
//
// PUBLIC
//
MS5611::MS5611(uint8_t CSn)
{
  _cspin = CSn;
  pinMode(_cspin, OUTPUT);
  digitalWrite(_cspin, HIGH);
  _temperature = -999;
  _pressure = -999;
  //init();
}

void MS5611::init()
{
  SPI.begin();
  reset();
  
  // Default values (C1 t/m C6 taken from example column in datasheet to test calculations):
  C[0] = 1;
  C[1] = 40127;
  C[2] = 36924;
  C[3] = 23317;
  C[4] = 23282;
  C[5] = 33464;
  C[6] = 28312;
  C[7] = 0xF0F0;
 
  // Read all factory parameters C0 to C7 from PROM
  SPI.beginTransaction(settingsA);
  for (uint8_t reg = 0; reg < 8; reg++)
  {
    C[reg] = readProm(reg);
  }
  SPI.endTransaction();
}

int MS5611::read(/*uint8_t bits*/)
{
  // VARIABLES NAMES BASED ON DATASHEET  <- Nice!
  convert(0x40, 12); // change 12 to bits if you want to go back to custom
  uint32_t D1 = readADC();

  convert(0x50, 9); // change 9 to bits if you want to go back to custom
  uint32_t D2 = readADC();

  // TODO the multiplications of these constants can be done in init()
  // but first they need to be verified.

  // TEMP & PRESS MATH - PAGE 7/20 of the datasheet
  //  - avoiding float type to make running on Tiny's not-impossible
  //  - running into issues with uint64_t so using uint32_t with adjustments
  //      (at the cost of reduced resolution for temperature).
  uint32_t Tref, dT;
  uint32_t dTC6;
  int32_t  TEMP;
  Tref = C[5] * 256UL;
  if (D2 < Tref ) {  // (to avoid signed int so we can bit-shift for divisioning)
    dT   = Tref - D2;
    dTC6  = ((uint64_t)dT * (uint64_t)C[6]) >> 23;
    TEMP = 2000 - dTC6;
  } else {
    dT   = D2 - Tref;
    dTC6  = ((uint64_t)dT * (uint64_t)C[6]) >> 23;
    TEMP = 2000 + dTC6;
  }

  // OFF = offT1 + TCO * dT = C2 * 2^16 + (C4 * dT ) / 2^7
  uint64_t offT1  =  (uint64_t)C[2] << 16;
  uint64_t TCOdT  = ((uint64_t)C[4] * (uint64_t)dT) >> 7;
  int64_t  OFF;
  if (D2 < Tref ) {
    OFF = offT1 - TCOdT;
  } else {
    OFF = offT1 + TCOdT;
  }

  // SENSE = sensT1 + TCS * dT = C1 * 2^15 + (C3 * dT ) / 2^8
  uint64_t sensT1 =  (uint64_t)C[1] << 15;
  uint64_t TCSdT  = ((uint64_t)C[3] * (uint64_t)dT) >> 8;
  int64_t  SENS;
  if (D2 < Tref ) {
    SENS   = sensT1 - TCSdT;
  } else {
    SENS   = sensT1 + TCSdT;
  }


  // SECOND ORDER COMPENSATION - PAGE 8/20 of the datasheet
  // COMMENT OUT < 2000 CORRECTION IF NOT NEEDED
  // NOTE TEMPERATURE IS IN 0.01 C
  uint32_t T2    = 0;
  uint32_t OFF2  = 0;
  uint32_t SENS2 = 0;
  if (TEMP < 2000)
  {
    uint64_t tSQ;
    T2     = ((uint64_t)dT * (uint64_t)dT) >> 31;
    tSQ    = (int32_t)TEMP - 2000L;
    tSQ   *= tSQ;
    OFF2   = (5ULL * (uint64_t)tSQ) >> 1;
    SENS2  = (5ULL * (uint64_t)tSQ) >> 2;
    // COMMENT OUT < -1500 CORRECTION IF NOT NEEDED
    if (TEMP < -1500)
    {
      tSQ    = (int32_t)TEMP - 2000L;
      tSQ   *= tSQ;
      OFF2  +=   7ULL * (uint64_t)tSQ;
      SENS2 += (11ULL * (uint64_t)tSQ) >> 1;
    }
  }

  TEMP -= T2;
  OFF  -= OFF2;
  SENS -= SENS2;
  //
  // END SECOND ORDER COMPENSATION
  //


  int64_t P  = (int64_t)D1 * SENS;
  P /= 2097152LL;
  P -= OFF;
  P /= 32768LL;

  _temperature = TEMP; 
  _pressure = (uint32_t)P; 

  return 0;
}


/////////////////////////////////////////////////////
//
// PRIVATE
//
void MS5611::reset()
{
  SPI.beginTransaction(settingsA);                      // start SPI transaction
  digitalWrite(_cspin, LOW);                            // pull CS line low
  SPI.transfer(0x1E);                                   // send reset command
  delay(4);
  digitalWrite(_cspin,HIGH);                            // pull CS line high
  SPI.endTransaction();                                 // end SPI transaction
}

void MS5611::convert(const uint8_t addr, uint8_t bits)
{
  uint8_t del[5] = {1, 2, 3, 5, 10};                    // array of MS5611 conversion time (in ms)
  bits = constrain(bits, 8, 12);
  uint8_t offset = (bits - 8) * 2;
  SPI.beginTransaction(settingsA);                      // start SPI transaction
  digitalWrite(_cspin, LOW);                            // pull CS line low
  SPI.transfer(addr + offset);                          // send command
  //delay(del[offset/2]);                                 // MS5611 needs some time for conversion; wait for this...
  wait(bits-7);                                           // improved delay times for faster speed
  digitalWrite(_cspin,HIGH);                            // pull CS line high
  SPI.endTransaction();                                 // end SPI transaction
}

void MS5611::wait(uint8_t OSR_mode){
  int DLY;  // Delay (microseconds)
  switch(OSR_mode){
    
    case 1://(OSR_256 @ 0.54ms)
      //OSR = 0x40;
      DLY = 540;
      break;
    
    case 2://(OSR_512 @ 1.06ms)
      //OSR = 0x42;
      DLY = 1060;
      break;
    
    case 3://(OSR_1024 @ 2.08ms)
      //OSR = 0x44;
      DLY = 2080;
      break;
    
    case 4://(OSR_2048 @ 4.13ms)
      //OSR = 0x46;
      DLY = 4130;
      break;

    //case 5: (the default case)
    
    default://(OSR_4096 @ 8.22ms)
      //OSR = 0x48;
      DLY = 8420;//8220; // NOT SURE WHY 8220 WONT WORK ANYMORE
      break;
  }
  delayMicroseconds(DLY);
}

uint16_t MS5611::readProm(uint8_t reg) {
  // read two bytes from SPI and return accumulated value
  reg = min(reg, 7);
  uint8_t offset = reg * 2;
  uint16_t val = 0;
  digitalWrite(_cspin, LOW);                            // pull CS line low
  SPI.transfer(0xA0 + offset);                          // send command
  val  = SPI.transfer(0x00) * 256;                      // read 8 bits of data (MSB)
  val += SPI.transfer(0x00);                            // read 8 bits of data (LSB)
  digitalWrite(_cspin,HIGH);                            // pull CS line high
  return val;
}

uint32_t MS5611::readADC() {
  // read three bytes from SPI and return accumulated value
  uint32_t val = 0UL;
  SPI.beginTransaction(settingsA);                      // start SPI transaction
  digitalWrite(_cspin, LOW);                            // pull CS line low
  SPI.transfer(0x00);                                   // send command
  val  = SPI.transfer(0x00) * 65536UL;                  // read 8 bits of data (MSB)
  val += SPI.transfer(0x00) * 256UL;                    // read 8 bits of data
  val += SPI.transfer(0x00);                            // read 8 bits of data (LSB)
  digitalWrite(_cspin,HIGH);                            // pull CS line high
  SPI.endTransaction();                                 // end SPI transaction
  return val;
}

// END OF FILE


vSpeed_mini/DEFAULT_SETTINGS.h
/**
 * FEEL FREE TO CHANGE ANY OF THE VALUES BELOW TO YOUR PREFERENCE.
 */


#ifndef DEFAULT_SETTINGS_H
#define DEFAULT_SETTINGS_H

class Default{
  public:
  
    bool ENABLE_CLIMB_BEEP           = 1;    // [1]YES, [0]NO
    bool ENABLE_SINK_BEEP            = 0;    // [1]YES, [0]NO
    bool ENABLE_SINK_ALARM           = 1;    // [1]YES, [0]NO
    int START_UP_VOLUME             = 10;    // Default 10 (0 to 31)
    int BEEP_TYPE                    = 1;    // DEFAULT [1]BASED_ON_VELOCITY, [2]BUFFERED_INCREMENTS
    
    float CLIMB_BEEP_TRIGGER       = 1.0;    // (Default 1.0 ft)
    float SINK_BEEP_TRIGGER       = -1.0;    // (Default -1.0 ft/s)
    float SINK_ALARM_TRIGGER     = -20.0;    // (Default -20.0 ft/s)

    float CLIMB_PITCH_MAX        = 600.0;    // (Default 600.0 Hz)
    float CLIMB_PITCH_MIN        = 400.0;    // (Default 400.0 Hz)
    
    float SINK_PITCH_MAX         = 250.0;    // (Default 250.0 Hz)
    float SINK_PITCH_MIN         = 150.0;    // (Default 150.0 Hz)
    float SINK_ALARM_PITCH      = 1000.0;    // (Default 1000.0 Hz)

    int BLUETOOTH_MODE               = 2;    // [0]DISABLE, [1]VSPEED_ANDROID, [2]FLYSKYHY_IOS, [3]XCTRACK_ANDROID

    int ALTITUDE_UNITS               = 1;    // [1]FEET, [2]METERS
    int VELOCITY_UNITS               = 1;    // [1]FEET_PER_SECOND, [2]METERS_PER_SECOND, [3]FEET_PER_MINUTE
};
    
#endif

